src/util/expr.h

```diff
@@ -161,6 +162,126 @@ class exprt:public irept
 
   void visit(class expr_visitort &visitor);
   void visit(class const_expr_visitort &visitor) const;
+
+  // Depth first search iterator - iterates over supplied expression
+  // and all its operands recursively.
+  template<typename expr_t>
+  class depth_iteratort final:
+  public std::iterator<std::forward_iterator_tag, expr_t, void>
+  {
+  public:
+    // Create end iterator
+    depth_iteratort() { }
+    // Create begin iterator, starting at the supplied node
+    explicit depth_iteratort(expr_t &root)
+    {
+      static_assert(std::is_same<exprt,
+      typename std::remove_const<expr_t>::type>::value,
+      "Only exprt and const exprt are allowed");
```

Comment:
This comment might be misplaced.

This check strips constness from the type and compares it against exprt - thus disallowing creation of depth_iterators of any other than exprt and const exprt. Changing remove_const to conditional would make the assert completely useless.